src/cmd/groff/contrib/groffer/perl/groffer.man
.TH GROFFER @MAN1EXT@ "@MDATE@" "Groff Version @VERSION@"
.SH NAME
groffer \- display groff files and man\~pages on X and tty
.
.SH "SYNOPSIS"
.\" The .SH was moved to this place in order to appease `apropos'.
.
.\" --------------------------------------------------------------------
.\" Legalize
.\" --------------------------------------------------------------------
.
.ig
groffer.1 - man page for groffer (section 1).

Source file position:  <groff_source_top>/contrib/groffer/perl/groffer.man
Installed position:    $prefix/share/man/man1/groffer.1

Last update: 13 Jun 2007
..
.de author
This file was written by Bernd Warken.
..
.de copyleft
Copyright (C) 2001,2002,2004,2005,2006 Free Software Foundation, Inc.
.
.P
This file is part of
.IR \%groffer ,
which is part of
.IR \%groff ,
a free software project.
.
You can redistribute it and/or modify it under the terms of the
.nh
.B GNU General Public License
.hy
as published by the
.nh
.BR "Free Software Foundation" ,
.hy
either version 2, or (at your option) any later version.
.
.P
You should have received a copy of the \f[CR]GNU General Public
License\f[] along with
.IR groff ,
see the files \%\f[CB]COPYING\f[] and \%\f[CB]LICENSE\f[] in the top
directory of the
.I groff
source package.
.
Or read the
.I man\~page
.BR gpl (1).
You can also write to the
.nh
.B Free Software Foundation, 51 Franklin St - Fifth Floor, Boston,
.BR "MA 02110-1301, USA" .
.hy
..
.
.\" --------------------------------------------------------------------
.\" Setup
.\" --------------------------------------------------------------------
.
.ds Ellipsis "\&.\|.\|.\""
.
.\" --------------------------------------------------------------------
.\" Macro definitions
.
.\" --------------------------------------------------------------------
.\" .CB  (<text>...)
.\"
.\" Print in constant-width bold font.
.\"
.de CB
.  ft CB
.  Text \\$*
.  ft
..
.\" --------------------------------------------------------------------
.\" .CI  (<text>...)
.\"
.\" Print in constant-width italic font.
.\"
.de CI
.  ft CI
.  Text \\$*
.  ft
..
.\" --------------------------------------------------------------------
.\" .CR  (<text>...)
.\"
.\" Print in constant-width roman font.
.\"
.de CR
.  ft CR
.  Text \\$*
.  ft
..
.\" --------------------------------------------------------------------
.\" .Error  (<text>...)
.\"
.\" Print error message to terminal and abort.
.\"
.de Error
.  tm \\$*
.  ab
..
.\" --------------------------------------------------------------------
.\" .Env_var  (<env_var_name> [<punct>])
.\"
.\" Display an environment variable, with optional punctuation.
.\"
.de Env_var
.  nh
.  SM
.  Text \f[CB]\\$1\f[]\\$2
.  hy
..
.\" --------------------------------------------------------------------
.\" .File_name  (<path_name>)
.\"
.\" Display a file or directory name in CB font.
.\"
.de File_name
.  Header_CB \\$@
..
.\" --------------------------------------------------------------------
.\" .Header_CB  (<path_name>)
.\"
.\" Display a line in CB font, for example after .TP
.\"
.de Header_CB
.  nh
.  Text \f[CB]\\$1\f[]\\$2
.  hy
..
.\"  --------------------------------------------------------------------
.\" .Text  (<text>...)
.\"
.\" Treat the arguments as text, no matter how they look.
.\"
.de Text
.  if \\n[.$]=0 \
.    return
.  nh
.  nop \)\\$*\)
.  hy
..
.\" --------------------------------------------------------------------
.\" .Topic  ([<indent>])
.\"
.\" A bulleted paragraph
.\"
.de Topic
.  ie \\n[.$]=0 \
.    ds @indent 2m\"
.  el \
.    ds @indent \\$1\"
.  TP \\*[@indent]
.  Text \[bu]
.  rm @indent
..

.\" End of macro definitions
.
.
.\" --------------------------------------------------------------------
.\" SH "SYNOPSIS"
.\" --------------------------------------------------------------------
.
.SY groffer
.OP option \*[Ellipsis]
.OP --
.OP \%filespec \*[Ellipsis]
.YS
.
.SY groffer
.BR -h | --help
.YS
.
.SY groffer
.BR -v | --version
.YS
.
.
.\" --------------------------------------------------------------------
.SH DESCRIPTION
.\" --------------------------------------------------------------------
.
The
.B \%groffer
program is the easiest way to use
.BR \%groff (@MAN1EXT@).
It can display arbitrary documents written in the
.I \%groff
language, see
.BR \%groff (@MAN7EXT@),
or other
.I \%roff
languages, see
.BR \%roff (@MAN7EXT@),
that are compatible to the original
.I \%troff
language.
.
It finds and runs all necessary
.I groff
preprocessors, such as
.BR @g@chem .
.
.
.P
The
.B \%groffer
program also includes many of the features for finding and displaying
the \%\f[CR]Unix\f[] manual pages
.nh
.RI ( man\~pages ),
.hy
such that it can be used as a replacement for a
.BR \%man (1)
program.
.
Moreover, compressed files that can be handled by
.BR \%gzip (1)
or
.BR \%bzip2 (1)
are decompressed on-the-fly.
.
.
.P
The normal usage is quite simple by supplying a file name or name of a
.I \%man\~page
without further options.
.
But the option handling has many possibilities for creating special
behaviors.
.
This can be done either in configuration files, with the shell
environment variable
.Env_var \%$GROFFER_OPT ,
or on the command line.
.
.
.P
The output can be generated and viewed in several different ways
available for
.IR \%groff .
.
This includes the
.I \%groff
native \%\f[CR]X\~Window\f[] viewer
.BR \%gxditview (@MAN1EXT@),
each
.IR \%Postcript ,
.IR \%pdf ,
or
.I \%dvi
display program, a web browser by generating
.I \%html
in
.IR \%www\~mode ,
or several
.I \%text\~modes
in text terminals.
.
.
.P
Most of the options that must be named when running
.B \%groff
directly are determined automatically for
.BR \%groffer ,
due to the internal usage of the
.BR \%grog (@MAN1EXT@)
program.
.
But all parts can also be controlled manually by arguments.
.
.
.P
Several file names can be specified on the command line arguments.
.
They are transformed into a single document in the normal way of
.BR \%groff .
.
.
.P
Option handling is done in \f[CR]GNU\f[] style.
.
Options and file names can be mixed freely.
.
The option
.RB ` \-\- '
closes the option handling, all following arguments are treated as
file names.
.
Long options can be abbreviated in several ways.
.
.
.\" --------------------------------------------------------------------
.SH "OPTION OVERVIEW"
.\" --------------------------------------------------------------------
.
.TP
.I breaking options
.RS
.P
.SY
.OP "-h~\fR|\fB\~--help"
.OP "-v~\fR|\fB\~--version"
.YS
.RE
.
.
.TP
.I \%groffer mode options
.RS
.P
.SY
.OP --auto
.OP --default
.OP --default\-modes mode1,mode2,\*[Ellipsis]
.OP --dvi
.OP --dvi\-viewer prog
.OP --groff
.OP --html
.OP --html\-viewer prog
.OP --mode display_mode
.OP --pdf
.OP --pdf\-viewer prog
.OP --ps
.OP --ps\-viewer prog
.OP --source
.OP --text
.OP --to\-stdout
.OP --tty
.OP --tty\-viewer prog
.OP --www
.OP --www\-viewer prog
.OP --x --X
.OP --x\-viewer --X\-viewer prog
.YS
.RE
.
.
.TP
.I options related to \%groff
.RS
.P
.OP -T~\fR|\fB\~--device device
.OP -Z~\fR|\fB\~--intermediate\-output~\fR|\fB\~--ditroff
.P
All further
.B \%groff
short options are accepted.
.RE
.
.
.TP
.I options for man\~pages
.SY
.OP --apropos
.OP --apropos\-data
.OP --apropos\-devel
.OP --apropos\-progs
.OP --man
.OP --no-man
.OP --no-special
.OP --whatis
.YS
.
.
.TP
.I long options taken over from GNU man
.RS
.P
.SY
.OP --all
.OP --ascii
.OP --ditroff
.OP --extension suffix
.OP --locale language
.OP --local-file
.OP "--location~\fR|\fB\~--where"
.OP --manpath dir1:dir2:\*[Ellipsis]
.OP --no-location
.OP --pager program
.OP --sections sec1:sec2:\*[Ellipsis]
.OP --systems sys1,sys2,\*[Ellipsis]
.OP --troff-device device
.YS
.P
Further long options of \f[CR]GNU\f[]
.B man
are accepted as well.
.RE
.
.
.TP
.I X Window Toolkit options
.RS
.P
.SY
.OP --bd~\fR|\fB\~--bordercolor pixels
.OP --bg~\fR|\fB\~--background color
.OP --bw~\fR|\fB\~--borderwidth pixels
.OP --display X-display
.OP --fg~\fR|\fB\~--foreground color
.OP --fn~\fR|\fB\~--ft~\fR|\fB\~--font font_name
.OP --geometry size_pos
.OP --resolution value
.OP --rv
.OP --title string
.OP --xrm X-resource
.YS
.RE
.
.
.TP
.I options for development
.RS
.P
.SY
.OP --debug
.OP --debug\-filenames
.OP --debug\-grog
.OP --debug\-keep
.OP --debug\-params
.OP --debug\-tmpdir
.OP --do\-nothing
.OP --print text
.OP -V
.YS
.RE
.
.
.TP
.I \%filespec arguments
.RS
.P
The
.I \%filespec
parameters are all arguments that are neither an option nor an option
argument.
.
They usually mean a file name or a
.I man page
searching scheme.
.
.
.P
In the following, the term
.I section_extension
is used.
.
It means a word that consists of a
.I man section
that is optionally followed by an
.IR extension .
.
The name of a
.I man section
is a single character from
.BR \%[1-9on] ,
the
.I extension
is some word.
.
The
.I extension
is mostly lacking.
.
.
.P
No
.I \%filespec
parameters means standard input.
.
.
.TP 10m
.B -
stands for standard input (can occur several times).
.
.
.TP
.I filename
the path name of an existing file.
.
.
.TP
.BI man: name ( section_extension )
.TQ
.BI man: name . section_extension
.TQ
.IB name ( section_extension )
.TQ
.IB name . section_extension
.TQ
.I "section_extension name"
search the \%man\~page
.I \%name
in the section with optional extension
.IR section_extension .
.
.
.TP
.BI man: name
\%man\~page in the lowest
.I \%man\~section
that has
.IR \%name .
.
.
.TP
.I name
if
.I \%name
is not an existing file search for the man\~page
.I \%name
in the lowest man\~section.
.
.RE
.
.
.\" --------------------------------------------------------------------
.SH "OPTION DETAILS"
.\" --------------------------------------------------------------------
.
The
.B \%groffer
program can usually be run with very few options.
.
But for special purposes, it supports many options.
.
These can be classified in 5 option classes.
.
.
.P
All short options of
.B \%groffer
are compatible with the short options of
.BR \%groff (@MAN1EXT@).
.
All long options of
.B \%groffer
are compatible with the long options of
.BR \%man (1).
.
.
.P
Arguments for long option names can be abbreviated in several ways.
.
First, the argument is checked whether it can be prolonged as is.
.
Furthermore, each minus sign
.B -
is considered as a starting point for a new abbreviation.
.
This leads to a set of multiple abbreviations for a single argument.
.
For example,
.B --de\-n\-f
can be used as an abbreviation for
.BR --debug\-not\-func ,
but
.B --de\-n
works as well.
.
If the abbreviation of the argument leads to several resulting options
an error is raised.
.
.
.P
These abbreviations are only allowed in the environment variable
.Env_var \%$GROFFER_OPT ,
but not in the configuration files.
.
In configuration, all long options must be exact.
.
.
.\" --------------------------------------------------------------------
.SS "groffer breaking Options"
.\" --------------------------------------------------------------------
.
As soon as one of these options is found on the command line it is
executed, printed to standard output, and the running
.B \%groffer
is terminated thereafter.
.
All other arguments are ignored.
.
.
.TP
.OP -h\fR|\fR--help
Print a helping information with a short explanation of option sto
standard output.
.
.
.TP
.OP -v\fR\fB--version
Print version information to standard output.
.
.
.\" --------------------------------------------------------------------
.SS "groffer Mode Options"
.\" --------------------------------------------------------------------
.
The display mode and the viewer programs are determined by these
options.
.
If none of these mode and viewer options is specified
.B \%groffer
tries to find a suitable display mode automatically.
.
The default modes are
.IR "mode pdf" ,
.IR "mode ps" ,
.IR "mode html" ,
.IR "mode x" ,
and
.I "mode dvi"
in \%\f[CR]X\~Window\f[] with different viewers and
.I mode tty
with device
.I latin1
under
.B less
on a terminal; other modes are tested if the programs for the main
default mode do not exist.
.
.
.P
In \%\f[CR]X\~Window\f[],
many programs create their own window when called.
.B \%groffer
can run these viewers as an independent program in the background.
.
As this does not work in text mode on a terminal (tty) there must be a
way to know which viewers are \%\f[CR]X\~Window\f[] graphical
programs.
.
The
.B \%groffer
script has a small set of information on some viewer names.
.
If a viewer argument of the command\-line chooses an element that is
kept as \%\f[CR]X\~Window\f[] program in this list it is treated as a
viewer that can run in the background.
.
All other, unknown viewer calls are not run in the background.
.
.
.P
For each mode, you are free to choose whatever viewer you want.
.
That need not be some graphical viewer suitable for this mode.
.
There is a chance to view the output source; for example, the
combination of the options
.B --mode=ps
and
.B --ps\-viewer=less
shows the content of the
.I Postscript
output, the source code, with the pager
.BR less .
.
.
.TP
.B --auto
Equivalent to
.BR --mode=auto .
.
.
.TP
.B --default
Reset all configuration from previously processed command line options
to the default values.
.
This is useful to wipe out all former options of the configuration, in
.Env_var \%$GROFFER_OPT ,
and restart option processing using only the rest of the command line.
.
.
.TP
.B --default\-modes mode1,mode2,\*[Ellipsis]
Set the sequence of modes for
.I \%auto\~mode
to the comma separated list given in the argument.
.
See
.B --mode
for details on modes.  Display in the default manner; actually, this
means to try the modes
.IR x ,
.IR ps ,
and
.I \%tty
in this sequence.
.
.
.
.TP
.B --dvi
Equivalent to
.BR --mode=\%dvi .
.
.
.TP
.B --dvi\-viewer prog
Choose a viewer program for
.IR \%dvi\~mode .
.
This can be a file name or a program to be searched in
.Env_var $PATH .
.
Known \%\f[CR]X\~Window\f[]
.I \%dvi
viewers include
.BR \%xdvi (1)
and
.BR \%dvilx (1)
.
In each case, arguments can be provided additionally.
.
.
.TP
.B --groff
Equivalent to
.BR --mode=groff .
.
.
.TP
.B --html
Equivalent to
.BR --mode=html .
.
.
.TP
.B --html\-viewer
Choose a web browser program for viewing in
.IR \%html\~mode .
.
It can be the path name of an executable file or a program in
.Env_var $PATH .
.
In each case, arguments can be provided additionally.
.
.
.TP
.BI --mode value
.
Set the display mode.
.
The following mode values are recognized:
.
.RS
.
.TP
.Header_CB auto
Select the automatic determination of the display mode.
.
The sequence of modes that are tried can be set with the
.B --default\-modes
option.
.
Useful for restoring the
.I \%default\~mode
when a different mode was specified before.
.
.
.TP
.Header_CB dvi
Display formatted input in a
.I \%dvi
viewer program.
.
By default, the formatted input is displayed with the
.BR \%xdvi (1)
program.
.BR --dvi .
.
.
.TP
.Header_CB groff
After the file determination, switch
.B \%groffer
to process the input like
.BR \%groff (@MAN1EXT@)
would do.
.
This disables the
.I \%groffer
viewing features.
.
.
.TP
.Header_CB html
Translate the input into html format and display the result in a web
browser program.
.
By default, the existence of a sequence of standard web browsers is
tested, starting with
.BR \%konqueror (1)
and
.BR \%mozilla (1).
The text html viewer is
.BR \%lynx (1).
.
.
.TP
.Header_CB pdf
Display formatted input in a
.I \%PDF
(Portable Document Format) viewer
program.
.
By default, the input is formatted by
.B \%groff
using the Postscript device, then it is transformed into the PDF file
format using
.BR \%gs (1),
or
.BR ps2pdf (1).
If that's not possible, the
.I Postscript mode (ps)
is used instead.
.
Finally it is displayed using different viewer programs.
.
.I \%pdf
has a big advantage because the text is displayed graphically and
is searchable as well.
.
.
.TP
.Header_CB ps
Display formatted input in a Postscript viewer program.
.
By default, the formatted input is displayed in one of many viewer
programs.
.
.
.TP
.Header_CB text
Format in a
.I \%groff\~text\~mode
and write the result to standard output without a pager or viewer
program.
.
The text device,
.I \%latin1
by default, can be chosen with option
.BR -T .
.
.
.TP
.Header_CB tty
Format in a
.I \%groff\~text\~mode
and write the result to standard output using a text pager program,
even when in \%\f[CR]X\~Window\f[].
.
.
.TP
.Header_CB www
Equivalent to
.BR --mode=html .
.
.
.TP
.Header_CB x
Display the formatted input in a native
.I roff
viewer.
.
By default, the formatted input is displayed with the
.BR \%gxditview (@MAN1EXT@)
program being distributed together with
.BR \%groff .
But the standard \%\f[CR]X\~Window\f[] tool
.BR \%xditview (1)
can also be chosen with the option
.B --x\-viewer .
The default resolution is
.BR 75\~dpi ,
but
.B 100\~dpi
are also possible.
.
The default
.I groff
device
for the resolution of
.B 75\~dpi
is
.BR X75\-12 ,
for
.B 100\~dpi
it is
.BR X100 .
.
The corresponding
.I "groff intermediate output"
for the actual device is generated and the result is displayed.
.
For a resolution of
.BR 100\~dpi ,
the default width of the geometry of the display program is chosen to
.BR 850\~dpi .
.
.
.TP
.Header_CB X
Equivalent to
.BR --mode=x .
.
.
.P
The following modes do not use the
.I \%groffer
viewing features.
.
They are only interesting for advanced applications.
.
.
.TP
.Header_CB groff
Generate device output with plain
.I \%groff
without using the special viewing features of
.IR \%groffer .
If no device was specified by option
.B -T
the
.I \%groff
default
.B \%ps
is assumed.
.
.
.TP
.Header_CB source
Output the roff source code of the input files without further
processing.
.
.
.RE
.
.
.TP
.B --pdf
Equivalent to
.BR --mode=pdf .
.
.
.TP
.B --pdf\-viewer prog
Choose a viewer program for
.IR \%pdf\~mode .
.
This can be a file name or a program to be searched in
.Env_var $PATH ;
arguments can be provided additionally.
.
.
.TP
.B --ps
Equivalent to
.BR --mode=ps .
.
.
.TP
.B --ps\-viewer prog
Choose a viewer program for
.IR \%ps\~mode .
.
This can be a file name or a program to be searched in
.Env_var $PATH .
.
Common Postscript viewers inlude
.BR \%gv (1),
.BR \%ghostview (1),
and
.BR \%gs (1),
.
In each case, arguments can be provided additionally.
.
.
.TP
.B --source
Equivalent
.BR --mode=source .
.
.
.TP
.B --text
Equivalent to
.BR --mode=text .
.
.
.TP
.B --to\-stdout
The file for the chosen mode is generated and its content is printed
to standard output.
.
It will not be displayed in graphical mode.
.
.
.TP
.B --tty
Equivalent to
.BR --mode=tty .
.
.
.TP
.B --tty\-viewer prog
Choose a text pager for mode
.IR tty .
The standard pager is
.BR less (1).
This option is eqivalent to
.I man
option
.BR --pager=prog .
The option argument can be a file name or a program to be searched in
.Env_var $PATH ;
arguments can be provided additionally.
.
.
.TP
.B --www
Equivalent to
.BR --mode=html .
.
.
.TP
.B --www\-viewer prog
Equivalent to
.B --html\-viewer .
.
.
.TP
.B --X~\fR|\fB\~--x
Equivalent to
.BR --mode=x .
.
.
.TP
.B --X\-viewer -- x\-viewer prog
Choose a viewer program for
.IR \%x\~mode .
Suitable viewer programs are
.BR \%gxditview (@MAN1EXT@)
which is the default and
.BR \%xditview (1).
The argument can be any executable file or a program in
.Env_var $PATH ;
arguments can be provided additionally.
.
.
.TP
.B --
Signals the end of option processing; all remaining arguments are
interpreted as
.I \%filespec
parameters.
.
.
.P
Besides these,
.B \%groffer
accepts all short options that are valid for the
.BR \%groff (@MAN1EXT@)
program.
.
All
.RB \%non- groffer
options are sent unmodified via
.B \%grog
to
.BR \%groff .
.
So postprocessors, macro packages, compatibility with
.I classical
.IR \%troff ,
and much more can be manually specified.
.
.
.\" --------------------------------------------------------------------
.SS "Options related to groff"
.\" --------------------------------------------------------------------
.
All short options of
.B \%groffer
are compatible with the short options of
.BR \%groff (@MAN1EXT@).
.
The following of
.B \%groff
options have either an additional special meaning within
.B \%groffer
or make sense for normal usage.
.
.
.P
Because of the special outputting behavior of the
.B \%groff
option
.B -Z
.B \%groffer
was designed to be switched into
.I \%groff\~mode ;
the
.I \%groffer
viewing features are disabled there.
.
The other
.B \%groff
options do not switch the mode, but allow to customize the formatting
process.
.
.
.TP
.B --a
This generates an ascii approximation of output in the
.IR \%text\~modes .
.
That could be important when the text pager has problems with control
sequences in
.IR "tty mode" .
.
.
.TP
.BI --m file
Add
.I \%file
as a
.I \%groff
macro file.
.
This is useful in case it cannot be recognized automatically.
.
.
.TP
.BI --P opt_or_arg
Send the argument
.I \%opt_or_arg
as an option or option argument to the actual
.B \%groff
postprocessor.
.
.
.TP
.B --T \fIdevname\fR ~\fR|\fB\~ --device \fIdevname\fR
.
This option determines
.BR \%groff 's
output device.
.
The most important devices are the text output devices for referring
to the different character sets, such as
.BR \%ascii ,
.BR \%utf8 ,
.BR \%latin1 ,
and others.
.
Each of these arguments switches
.B \%groffer
into a
.I \%text\~mode
using this device, to
.I \%mode\~tty
if the actual mode is not a
.IR \%text\~mode .
.
The following
.I \%devname
arguments are mapped to the corresponding
.B \%groffer
.B --mode=\fIdevname\fR
option:
.BR \%dvi ,
.BR \%html ,
and
.BR \%ps .
All
.B \%X*
arguments are mapped to
.IR \%mode\~x .
Each other
.I \%devname
argument switches to
.I \%mode\~groff
using this device.
.
.
.TP
.B --X
is equivalent to
.BR "groff \-X" .
It displays the
.I groff intermediate output
with
.BR gxditview .
As the quality is relatively bad this option is deprecated; use
.B --X
instead because the
.I \%x\~mode
uses an
.IR X *
device for a better display.
.
.
.TP
.B -Z~\fR|\fB\~--intermediate-output~\fR|\fB\~--ditroff
Switch into
.I \%groff\~mode
and format the input with the
.I \%groff intermediate output
without postprocessing; see
.BR \%groff_out (@MAN5EXT@).
This is equivalent to option
.B --ditroff
of
.IR \%man ,
which can be used as well.
.
.
.P
All other
.B \%groff
options are supported by
.BR \%groffer ,
but they are just transparently transferred to
.B \%groff
without any intervention.
.
The options that are not explicitly handled by
.B \%groffer
are transparently passed to
.BR \%groff .
.
Therefore these transparent options are not documented here, but in
.BR \%groff (@MAN1EXT@).
Due to the automatism in
.BR \%groffer ,
none of these
.B \%groff
options should be needed, except for advanced usage.
.
.
.\" --------------------------------------------------------------------
.SS "Options for man\~pages"
.\" --------------------------------------------------------------------
.
.TP
.B --apropos
Start the
.BR \%apropos (1)
command or facility of
.BR \%man (1)
for searching the
.I \%filespec
arguments within all
.I \%man\~page
descriptions.
.
Each
.I \%filespec
argument is taken for search as it is;
.I section
specific parts are not handled, such that
.B 7 groff
searches for the two arguments
.B 7
and
.BR groff ,
with a large result; for the
.I \%filespec
.B groff.7
nothing will be found.
.
The
.I language
locale is handled only when the called programs do support this; the
GNU
.B apropos
and

src/cmd/groff/contrib/groffer/shell/groffer.sh
#! /bin/sh

# groffer - display groff files

# Source file position: <groff-source>/contrib/groffer/groffer.sh
# Installed position: <prefix>/bin/groffer

# Copyright (C) 2001,2002,2003,2004,2005,2006
# Free Software Foundation, Inc.
# Written by Bernd Warken

# Last update: 22 Oct 2006

# This file is part of `groffer', which is part of `groff'.

# `groff' is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.

# `groff' is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with `groff'; see the files COPYING and LICENSE in the top
# directory of the `groff' source.  If not, write to the Free Software
# Foundation, 51 Franklin St - Fifth Floor, Boston, MA 02110-1301,
# USA.

########################################################################

export GROFFER_OPT;		# option environment for groffer

export _CONF_FILE_ETC;		# configuration file in /etc
export _CONF_FILE_HOME;		# configuration file in $HOME
export _CONF_FILES;		# configuration files
_CONF_FILE_ETC='/etc/groff/groffer.conf';
_CONF_FILE_HOME="${HOME}/.groff/groffer.conf";
_CONF_FILES="${_CONF_FILE_ETC} ${_CONF_FILE_HOME}";

# characters

export _AT;
export _SP;
export _SQ;
export _TAB;

_AT='@';
_SP=' ';
_SQ="'";
_TAB='	';

export _ERROR;
_ERROR='7';			# for syntax errors; no `-1' in `ash'

# @...@ constructs

export _BEFORE_MAKE;
if test _@VERSION@_ = _${_AT}VERSION${_AT}_
then
  # script before `make'
  _BEFORE_MAKE='yes';
else
  _BEFORE_MAKE='no';
fi;

export _AT_BINDIR_AT;
export _AT_G_AT;
export _AT_LIBDIR_AT;
export _GROFFER_LIBDIR;
case "${_BEFORE_MAKE}" in
yes)
  self="$0";
  case "${self}" in
  /*) :; ;;
  *)
    curdir="$(pwd)";
    case "${curdir}" in
    */)
      self="${curdir}${self}";
      ;;
    *)
      self="${curdir}/${self}";
      ;;
    esac;
    ;;
  esac;
  groffer_shell_dir="$(dirname ${self})";
  case "${groffer_shell_dir}" in
  */) :; ;;
  *) groffer_shell_dir="${groffer_shell_dir}/";
  esac;
  groffer_top_dir="${groffer_shell_dir}../";
  _AT_G_AT='';
  _AT_BINDIR_AT="${groffer_shell_dir}";
  _AT_LIBDIR_AT="${groffer_shell_dir}";
  _GROFFER_LIBDIR="${_AT_LIBDIR_AT}";
  _VERSION_SH="${groffer_top_dir}version.sh";
  ;;
no)
  _AT_G_AT='@g@';
  _AT_BINDIR_AT='@BINDIR@';
  case "${_AT_BINDIR_AT}" in
  */) :; ;;
  *) _AT_BINDIR_AT="${_AT_BINDIR_AT}/";
  esac;
  _AT_LIBDIR_AT='@libdir@';
  case "${_AT_LIBDIR_AT}" in
  */) :; ;;
  *) _AT_LIBDIR_AT="${_AT_LIBDIR_AT}/";
  esac;
  _GROFFER_LIBDIR="${_AT_LIBDIR_AT}"'groff/groffer/';
  _VERSION_SH="${_GROFFER_LIBDIR}"'version.sh';
  ;;
esac;

if test -f "${_VERSION_SH}"
then
  . "${_VERSION_SH}";
fi;

export _GROFF_VERSION;
case "${_BEFORE_MAKE}" in
yes)
  _GROFF_VERSION="${_GROFF_VERSION_PRESET}";
  ;;
no)
  _GROFF_VERSION='@VERSION@';
  ;;
esac;

export _GROFFER2_SH;		# file name of the script that follows up
_GROFFER2_SH="${_GROFFER_LIBDIR}"'groffer2.sh';

export _GROFFER_SH;		# file name of this shell script
case "$0" in
*groffer*)
  _GROFFER_SH="$0";
  ;;
*)
  echo 'The groffer script should be started directly.' >&2
  exit 1;
  ;;
esac;

export _NULL_DEV;
if test -c /dev/null
then
  _NULL_DEV="/dev/null";
else
  _NULL_DEV="NUL";
fi;


# Test of the `$()' construct.
if test _"$(echo "$(echo 'test')")"_ \
     != _test_
then
  echo 'The "$()" construct did not work.' >&2;
  exit "${_ERROR}";
fi;

# Test of sed program
if test _"$(echo red | sed 's/r/s/')"_ != _sed_
then
  echo 'The sed program did not work.' >&2;
  exit "${_ERROR}";
fi;

# for some systems it is necessary to set some unset variables to `C'
# according to Autobook, ch. 22
for var in LANG LC_ALL LC_MESSAGES LC_CTYPES LANGUAGES
do
  if eval test _"\${$var+set}"_ = _set_
  then
    eval ${var}='C';
    eval export ${var};
  fi;
done;


########################### configuration

# read and transform the configuration files, execute the arising commands
for f in "${_CONF_FILE_HOME}" "${_CONF_FILE_ETC}"
do
  if test -f "$f"
  then
    o="";			# $o means groffer option
    # use "" quotes because of ksh and posh
    eval "$(cat "$f" | sed -n '
# Ignore comments
/^['"${_SP}${_TAB}"']*#/d
# Delete leading and final space
s/^['"${_SP}${_TAB}"']*//
s/['"${_SP}${_TAB}"']*$//
# Print all lines with shell commands, those not starting with -
/^[^-]/p
# Remove all single and double quotes
s/['"${_SQ}"'"]//g
# Replace empty arguments
s/^\(-[^ ]*\)=$/o="${o} \1 '"${_SQ}${_SQ}"'"/p
# Replace division between option and argument by single space
s/[='"${_SP}${_TAB}"']['"${_SP}${_TAB}"']*/'"${_SP}"'/
# Handle lines without spaces
s/^\(-[^'"${_SP}"']*\)$/o="${o} \1"/p
# Encircle the remaining arguments with single quotes
s/^\(-[^ ]*\) \(.*\)$/o="${o} \1 '"${_SQ}"'\2'"${_SQ}"'"/p
')"

    # Remove leading space
    o="$(echo "$o" | sed 's/^ *//')";
    if test _"${o}"_ != __
    then
      if test _"{GROFFER_OPT}"_ = __
      then
        GROFFER_OPT="${o}";
      else
        GROFFER_OPT="${o} ${GROFFER_OPT}";
      fi;
    fi;
  fi;
done;

# integrate $GROFFER_OPT into the command line; it isn't needed any more
if test _"${GROFFER_OPT}"_ != __
then
  eval set x "${GROFFER_OPT}" '"$@"';
  shift;
  GROFFER_OPT='';
fi;


########################### Determine the shell

export _SHELL;

supports_func=no;
foo() { echo bar; } 2>${_NULL_DEV};
if test _"$(foo)"_ = _bar_
then
  supports_func=yes;
fi;

# use "``" instead of "$()" for using the case ")" construct
# do not use "" quotes because of ksh
_SHELL=`
  # $x means list.
  # $s means shell.
  # The command line arguments are taken over.
  # Shifting herein does not have an effect outside.
  export x;
  case " $*" in
  *\ --sh*)			# abbreviation for --shell
    x='';
    s='';
    # determine all --shell arguments, store them in $x in reverse order
    while test $# != 0
    do
      case "$1" in
      --shell|--sh|--she|--shel)
        if test "$#" -ge 2
        then
          s="$2";
          shift;
        fi;
        ;;
      --shell=*|--sh=*|--she=*|--shel=*)
        # delete up to first "=" character
        s="$(echo x"$1" | sed 's/^x[^=]*=//')";
        ;;
      *)
        shift;
        continue;
      esac;
      if test _"${x}"_ = __
      then
        x="'${s}'";
      else
        x="'${s}' ${x}";
      fi;
      shift;
    done;

    # from all possible shells in $x determine the first being a shell
    # or being empty
    s="$(
      # "" quotes because of posh
      eval set x "${x}";
      shift;
      if test $# != 0
      then
        for i
        do
          if test _"$i"_ = __
          then
            if test _"${supports_func}"_ = _yes_
            then
              # use the empty argument as the default shell
              echo 'standard shell';
              break;
            else
              echo groffer: standard shell does not support functions. >&2;
              continue;
            fi;
          else
            # test $i on being a shell program;
            # use this kind of quoting for posh
            if test _"$(eval "$i -c 'echo ok'" 2>${_NULL_DEV})"_ = _ok_ >&2
            then
              # test whether shell supports functions
              if eval "$i -c 'foo () { exit 0; }; foo'" 2>${_NULL_DEV}
              then
                # shell with function support found
                cat <<EOF
${i}
EOF
                break;
              else
                # if not being a shell with function support go on searching
                echo groffer: argument $i is not a shell \
with function support. >&2
                continue;
              fi;
            else
              # if not being a shell go on searching
              echo groffer: argument $i is not a shell. >&2
              continue;
            fi;
          fi;
        done;
      fi;
    )";
    if test _"${s}"_ != __
    then
      cat <<EOF
${s}
EOF
    fi;
    ;;
  esac;
`

########################### test fast shells for automatic run

if test _"${_SHELL}"_ = __
then
  # shell sorted by speed, bash is very slow
  for s in ksh ash dash pdksh zsh posh sh bash
  do
    # test on shell with function support
    if eval "$s -c 'foo () { exit 0; }; foo'" 2>${_NULL_DEV}
    then
      _SHELL="$s";
      break;
    fi;
  done;
fi;


########################### start groffer2.sh

if test _"${_SHELL}"_ = _'standard shell'_
then
  _SHELL='';
fi;

if test _"${_SHELL}"_ = __
then
  # no shell found, so start groffer2.sh normally
  if test _${supports_func}_ = _yes_
  then
    eval . "'${_GROFFER2_SH}'" '"$@"';
    exit;
  else
    echo groffer: standard shell does not support functions; no shell works.\
Get some free working shell such as bash. >&2
    exit "${_ERROR}";
  fi;
else
  # start groffer2.sh with the found $_SHELL
  # do not quote $_SHELL to allow arguments
  eval exec "${_SHELL} '${_GROFFER2_SH}'" '"$@"';
  exit;
fi;
